metagraph/src/graph/annotated_graph_algorithm.cpp

```diff
@@ -15,249 +12,416 @@ namespace graph {
-        // reduce calls to the annotator
-        if (labels_in_infrequent.size() || labels_out_infrequent.size()) {
-            sdsl::int_vector<> counts = construct_diff_label_count_vector(
-                anno_graph, labels_in_infrequent, labels_out_infrequent, num_threads
-            );
-
-            // the width of counts is double, since it's both in and out counts interleaved
-            uint32_t width = counts.width() >> 1;
-            uint64_t int_mask = (uint64_t(1) << width) - 1;
-
-            // Flatten count vector to bitmap if all labels were infrequent
-            if (labels_in_frequent.empty() && labels_out_frequent.empty()) {
-                sdsl::bit_vector mask(anno_graph.get_graph().max_index() + 1, false);
-
-                call_nonzeros(counts, [&](DeBruijnGraph::node_index i, size_t count) {
-                    if (i != DeBruijnGraph::npos
-                            && is_node_in_mask(i, [&]() { return count & int_mask; },
-                                                  [&]() { return count >> width; }))
-                        mask[i] = true;
-                });
-
-                return std::make_unique<bitmap_vector>(std::move(mask));
-            }
+    }, num_threads);
+    std::atomic_thread_fence(std::memory_order_acquire);
+
+    masked_graph.set_mask(new bitmap_vector(std::move(mask)));
 
-            // If at least one of the labels was frequent, construct a lazy bitmap
-            // which references both the Annotator and the count vector
-            auto count_frequent_in_labels = build_label_counter(anno_graph,
-                                                                labels_in_frequent);
-            auto count_frequent_out_labels = build_label_counter(anno_graph,
-                                                                 labels_out_frequent);
-            return std::make_unique<bitmap_lazy>([=](uint64_t i) {
-                auto count = counts[i];
-                return i != DeBruijnGraph::npos
-                    && is_node_in_mask(i,
-                            [&]() { return (count & int_mask) + count_frequent_in_labels(i); },
-                            [&]() { return (count >> width) + count_frequent_out_labels(i); });
-            }, counts.size());
+    logger->trace("Kept {} out of {} unitigs with average length {}",
+                  kept_unitigs, total_unitigs,
+                  static_cast<double>(num_kept_nodes + kept_unitigs * (masked_graph.get_k() - 1))
+                      / kept_unitigs);
+}
+
+void update_masked_graph_by_node(MaskedDeBruijnGraph &masked_graph,
+                                 const KeepNode &keep_node,
+                                 size_t num_threads) {
+    bool parallel = num_threads > 1;
+    std::atomic<size_t> kept_nodes(0);
+    size_t total_nodes = masked_graph.num_nodes();
+
+    const auto &in_mask = dynamic_cast<const bitmap_vector&>(masked_graph.get_mask()).data();
+    sdsl::bit_vector mask(in_mask);
```

Comment:
```suggestion
    sdsl::bit_vector mask = in_mask;
```

Reply:
Use copy assignment operators to make it clear that no conversion is happening and this just creates a copy.